### PR TITLE
refactor: handle trainee updates in integration service

### PR DESCRIPTION
**Pull request by CaiWillis**
TIS21-2873:[Filter Recommendations with ES] - Event Handling Job For Data Changes (TIS)

### CODE REVIEW

**Review by JosephKelly — APPROVED** [review]

File: src/test/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/service/CdcTraineeUpdateServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  @Mock
+  CdcMessagePublisher publisher;
+
+  @Spy
+  MasterDoctorViewMapper mapper;
+
+  private MasterDoctorView masterDoctorView = CdcTestDataGenerator.getTestMasterDoctorView();
+
+  void shouldAddNewFields() {
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/service/CdcTraineeUpdateServiceTest.java:62]: Looks like it's missing a `@Test`.

File: src/test/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/service/CdcTraineeUpdateServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    when(repository.findByGmcReferenceNumber(any())).thenReturn(List.of(masterDoctorView));
+
+    var traineeUpdates =
+        ConnectionInfoDto.builder().build();
+    cdcTraineeUpdateService.addNewEntity(traineeUpdates);
+
+    verify(repository).save(any());
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/service/CdcTraineeUpdateServiceTest.java:69]: I can't see assertions that line up with the method name.  You could use a `@Captor MasterDoctorView masterDoctorView;` to check the fields are added/removed as expected.

**Review by JosephKelly — CHANGES_REQUESTED** [review]
Also pushed a branch with a tiny tweak to improve code coverage.

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/service/CdcTraineeUpdateService.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+   *
+   * @param entity trainee info to add to index
+   */
+  @Override
+  public void upsertEntity(ConnectionInfoDto entity) {
+    final var repository = getRepository();
+    final var existingView = repository.findByGmcReferenceNumber(entity.getGmcReferenceNumber());
+    if (existingView.isEmpty()) {
+      upsertTraineeInfoWithNullGmcNumber(entity);
+    } else {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/service/CdcTraineeUpdateService.java:59]: It's not obvious that we should / shouldn't publish the "updatedView" even if there is no GMC information.  I think the tests should clearly `verifyNoInteractions()` or verify the expected CdcMessagePublisher invocation(s)

**Review by JosephKelly — COMMENTED** [review]
Sorry I left a partial review to get stale.  I've left a little comment and had been working on a slight refactoring to the `CdcTraineeUpdateService.upsertEntity()` method... it's probably different enough to raise after this has been fully reviewed/merged.

I've left a comment by part of the implementation that seems to be a bit more "unspecified" behavior.  Let me know if that needs clarifying

File: src/test/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/service/CdcTraineeUpdateServiceTest.java (hunk 2 above)
Comment (JosephKelly) [reply src/test/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/service/CdcTraineeUpdateServiceTest.java]: I could be tired but I have inferred a reliance on GMC number being null when we have never received a GMC record but this isn't verified by an assertion.

**Review by JosephKelly — APPROVED** [review]

**Review by hihilary — APPROVED** [review]

**Pull request merged**